### ISSUE TITLE
SQLite dialect does not produce a proper description of a duplicate column

### ISSUE THREAD

**Issue by vmarkovtsev** [sqlalchemy/sqlalchemy#7504]
### Describe the bug

See the reproduction source code. I select two same columns in Core. Result column descriptions are drastically different when compiled with SQLite dialect. That leads to crashes in the downstream projects that rely on `_result_columns`, for example, [encode/databases](https://github.com/encode/databases/blob/master/databases/backends/sqlite.py#L180). The subsequent crash is
```
    metadata = CursorResultMetaData(context, cursor.description)
/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/cursor.py:180: in __init__
    raw = self._merge_cursor_description(
/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/cursor.py:415: in _merge_cursor_description
    return [
/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/cursor.py:415: in <listcomp>
    return [
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sqlalchemy.engine.cursor.CursorResultMetaData object at 0x7f2e0195bee0>
context = <sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x7f2e01928e50>
cursor_description = (('pr_node_id', None, None, None, None, None, ...), ('repository_full_name', None, None, None, None, None, ...), ('rel...e, None, None, ...), ('author', None, None, None, None, None, ...), ('merger', None, None, None, None, None, ...), ...)
result_columns = [('pr_node_id', 'pr_node_id', (Column('pr_node_id', BigInteger(), table=<done_pull_request_facts>, primary_key=True, n...(), table=<done_pull_request_facts>), 'merger', 'merger', 'github.done_pull_request_facts_merger'), BigInteger()), ...]

    def _merge_textual_cols_by_position(
        self, context, cursor_description, result_columns
    ):
        num_ctx_cols = len(result_columns) if result_columns else None
    
        if num_ctx_cols > len(cursor_description):
            util.warn(
                "Number of columns in textual SQL (%d) is "
                "smaller than number of columns requested (%d)"
                % (num_ctx_cols, len(cursor_description))
            )
        seen = set()
        for (
            idx,
            colname,
            untranslated,
            coltype,
        ) in self._colnames_from_description(context, cursor_description):
            if idx < num_ctx_cols:
                ctx_rec = result_columns[idx]
                obj = ctx_rec[RM_OBJECTS]
                ridx = idx
                mapped_type = ctx_rec[RM_TYPE]
>               if obj[0] in seen:
E               IndexError: tuple index out of range

/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/cursor.py:503: IndexError
```

### To Reproduce

```python
from sqlalchemy import Column, Integer, select
from sqlalchemy.dialects.sqlite import pysqlite
from sqlalchemy.engine import ExecutionContext
from sqlalchemy.ext.declarative import declarative_base


class Foo(declarative_base()):
    __tablename__ = "whatever"

    bar = Column(Integer(), primary_key=True)


class CompilationContext:
    def __init__(self, context: ExecutionContext):
        self.context = context


def main():
    query = select([Foo.bar, Foo.bar])
    compiled = query.compile(dialect=pysqlite.dialect(paramstyle="qmark"))
    print(compiled._result_columns[0])
    print(compiled._result_columns[1])


if __name__ == "__main__":
    main()
```

### Error

The provided code prints:
```
('bar', 'bar', (Column('bar', Integer(), table=<whatever>, primary_key=True, nullable=False), 'bar', 'bar', 'whatever_bar'), Integer())
('bar__1', '%(575564119221534721 bar_)s', (), Integer())
``` 
The third tuple element in the second item is an empty tuple instead of a proper pointer to Column.

### Versions

- OS: Ubuntu 20.04
- Python: 3.8.10
- SQLAlchemy: 1.4.29
- Database: SQLite
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pysqlite

### Additional context

This code works differently on 1.3.x, it prints:
```
('bar', 'bar', (Column('bar', Integer(), table=<whatever>, primary_key=True, nullable=False), 'bar', 'bar'), Integer())
Traceback (most recent call last):
  File "reproduce.py", line 26, in <module>
    main()
  File "reproduce.py", line 22, in main
    print(compiled._result_columns[1])
IndexError: list index out of range
```
In other words, 1.3.x does not generate the second element for the duplicate.

**Comment by CaselIT** [comment]
Hi,

I'm not sure what the issue is, the query is correctly executed in the db and the result returned.
```py
import sqlalchemy as sa
from sqlalchemy.orm import Session,declarative_base

Base = declarative_base()


class Foo(Base):
    __tablename__ = "whatever"

    bar = sa.Column(sa.Integer(), primary_key=True)

def main():
    e = sa.create_engine("sqlite:///")
    query = sa.select([Foo.bar, Foo.bar])
    compiled = query.compile(e)
    print(compiled._result_columns[0])
    print(compiled._result_columns[1])
    Base.metadata.create_all(e)
    with Session(e) as s:
        s.add_all([Foo(), Foo(), Foo()])
        s.commit()
    with e.connect() as c:
        print(c.execute(query).all())


if __name__ == "__main__":
    main()
```

This will print  `[(1, 1), (2, 2), (3, 3)]` as expected

**Comment by vmarkovtsev** [comment]
The downstream library needs to compile the Core query to string + list of arguments and fetch the results with a foreign aiosqlite cursor. I doubt such a public API is possible.

Anyway, I got the point, will have to read the new 1.4 code and figure out how the heck it works with those `_result_columns`.

**Comment by CaselIT** [comment]
> Anyway, I got the point, will have to read the new 1.4 code and figure out how the heck it works with those `_result_columns`.

It was not my intention to dismiss your question. If you can explain what api you need for the downstream library we can try to accomodate it.

I think this is caused by the change that allows duplicates columns starting from 1.4 https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#change-4753

**Comment by CaselIT** [comment]
Reading https://github.com/encode/databases/issues/439 I really don't think I've dismissed anything.

The bug here is regarding a private api. I just replied that there is no expectation that private api keeps working the same between releases.

If you would like to continue the discussion on a constructive way feel free to reply.

In any case happy holidays

**Comment by zzzeek** [comment]
> The downstream library needs to compile the Core query to string + list of arguments and fetch the results with a foreign aiosqlite cursor. I doubt such a public API is possible.
> 
> Anyway, I got the point, will have to read the new 1.4 code and figure out how the heck it works with those `_result_columns`.

> This is to mirror the [upstream issue in SQLAlchemy 1.4](https://github.com/sqlalchemy/sqlalchemy/issues/7504). Those folks happily discard our problem because we are using the private API. As if we had a choice. 

hey there, I also just read your comment at https://github.com/encode/databases/issues/439   Why do you feel that you "have no choice" but to use SQLAlchemy's private API?   Just to be clear, the typical course of action is to create a feature request with us that describes your use case such that we may consider the feasibility of providing stable, public API to satisfy a particular use case.  We do this all the time, and while results are not guaranteed or necessarily immediate, would provide a better long term outcome if we are indeed able to accommodate what it is you need.  

in this case, the public API retains the duplicate columns which as described at https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#select-objects-and-derived-from-clauses-allow-for-duplicate-columns-and-column-labels is now supported.  so you can get those columns straight from the `.selected_column` collection:

```py
def main():
    query = select([Foo.bar, Foo.bar])

    print(query.selected_columns)
```


output:

    ImmutableColumnCollection(whatever.bar, whatever.bar)

**Comment by vmarkovtsev** [comment]
Thank you all, `.selected_columns` seem very helpful indeed, I will give it a shot!
